drivers/video/video_arducam_mega.c

```diff
@@ -0,0 +1,1397 @@
+#define CAM_REG_MANUAL_EXPOSURE_BIT_15_8      0X34
+#define CAM_REG_MANUAL_EXPOSURE_BIT_7_0       0X35
+#define CAM_REG_BURST_FIFO_READ_OPERATION     0X3C
+#define CAM_REG_SINGLE_FIFO_READ_OPERATION    0X3D
+#define CAM_REG_SENSOR_ID                     0x40
+#define CAM_REG_YEAR_SDK                      0x41
+#define CAM_REG_MONTH_SDK                     0x42
+#define CAM_REG_DAY_SDK                       0x43
+#define CAM_REG_SENSOR_STATE                  0x44
+#define CAM_REG_FPGA_VERSION_NUMBER           0x49
+#define CAM_REG_DEBUG_DEVICE_ADDRESS          0X0A
+#define CAM_REG_DEBUG_REGISTER_HIGH           0X0B
+#define CAM_REG_DEBUG_REGISTER_LOW            0X0C
+#define CAM_REG_DEBUG_REGISTER_VALUE          0X0D
+
+#define SENSOR_STATE_IDLE   (1 << 1)
+#define SENSOR_RESET_ENABLE (1 << 6)
+
+#define CTR_WHITEBALANCE 0X02
+#define CTR_EXPOSURE     0X01
+#define CTR_GAIN         0X00
+
+#define AC_STACK_SIZE 4096
+#define AC_PRIORITY   5
+
+#define ARDUCAM_MEGA_CAPTURE_TRIES 200
+
+K_THREAD_STACK_DEFINE(ac_stack_area, AC_STACK_SIZE);
+
+struct k_work_q ac_work_q;
+
+struct arducam_mega_config {
+	struct spi_dt_spec bus;
+};
+
+struct arducam_mega_ctrls {
+	struct video_ctrl reset;
+	struct video_ctrl brightness;
+	struct video_ctrl contrast;
+	struct video_ctrl saturation;
+	struct video_ctrl ev;
+	struct video_ctrl whitebal;
+	struct video_ctrl colorfx;
+	struct video_ctrl quality;
+	struct video_ctrl lowpower;
+	struct video_ctrl whitebalauto;
+	struct video_ctrl sharpness;
+	struct {
+		struct video_ctrl exp_auto;
+		struct video_ctrl exposure;
+	};
+	struct {
+		struct video_ctrl gain;
+		struct video_ctrl gain_auto;
+	};
+	struct video_ctrl focus_auto;
+	/* Read only registers */
+	struct video_ctrl support_resolution;
+	struct video_ctrl support_special_effects;
+	struct video_ctrl linkfreq;
+};
+
+struct arducam_mega_data {
+	struct arducam_mega_ctrls ctrls;
+	struct video_format fmt;
+
+	const struct device *dev;
+	struct k_fifo fifo_in;
+	struct k_fifo fifo_out;
+	struct k_work buf_work;
+	struct k_timer stream_schedule_timer;
+	struct k_poll_signal *signal;
+	uint8_t fifo_first_read;
+	uint32_t fifo_length;
+	uint8_t stream_on;
+	uint32_t features;
+	uint32_t camera_id;
+};
+
+#define ARDUCAM_MEGA_VIDEO_FORMAT_CAP(width, height, format)                                       \
+	{.pixelformat = (format),                                                                  \
+	 .width_min = (width),                                                                     \
+	 .width_max = (width),                                                                     \
+	 .height_min = (height),                                                                   \
+	 .height_max = (height),                                                                   \
+	 .width_step = 0,                                                                          \
+	 .height_step = 0}
+
+static struct video_format_cap fmts[] = {
```

Comment:
I realize that because this table is global, this means multiple instance of different arducam mega on a same system will be problematic...

But the current format definition is not ideal, and a future API will make these kinds of things easier, so possible to let it like it is now in the meantime.